ops/pebble.py

```diff
@@ -685,6 +747,226 @@ def __repr__(self):
+        if timeout is not None:
+            # A bit more than the command timeout to ensure that happens first
+            timeout += 1
+        change = self._client.wait_change(self._change_id, timeout=timeout)
+
+        # If stdin reader thread is running, stop it
+        if self._cancel_stdin is not None:
+            self._cancel_stdin()
+
+        # Wait for all threads to finish (e.g., message barrier sent)
+        for thread in self._threads:
+            thread.join()
+
+        if change.err:
+            raise ChangeError(change.err, change)
+        exit_code = change.data.get('exit-code', -1)
+        return exit_code
+
+    def wait_output(self) -> typing.Tuple[typing.AnyStr, typing.AnyStr]:
+        """Wait for the process to finish and return tuple of (stdout, stderr).
+
+        If a timeout was specified to the :meth:`Client.exec` call, this waits
+        at most that duration. If combine_stderr was True, stdout will include
+        the process's standard error, and stderr will be None.
+
+        Raises:
+            ExecError: if the process exits with a non-zero exit code.
+        """
+        if self._encoding is not None:
+            out = io.StringIO()
+            err = io.StringIO()
+        else:
+            out = io.BytesIO()
+            err = io.BytesIO()
+
+        t = _start_thread(shutil.copyfileobj, self.stdout, out)
+        self._threads.append(t)
+        t = _start_thread(shutil.copyfileobj, self.stderr, err)
+        self._threads.append(t)
+
+        exit_code = self._wait()
+        if exit_code != 0:
+            raise ExecError(self._command, exit_code, out.getvalue(), err.getvalue())
+
+        return (out.getvalue(), err.getvalue())
+
+    def send_signal(self, signum: int):
+        """Send the given signal number to the running process.
+
+        Args:
+            signum: Number of signal to send, e.g., 1 or :data:`signal.SIGHUP`.
+        """
+        msg = json.dumps({'command': 'signal', 'signal': signum}, sort_keys=True)
+        self._control_ws.send(msg)
+
+
+def _has_fileno(f):
+    """Return True if the file-like object has a valid fileno() method."""
+    try:
+        f.fileno()
+        return True
+    except Exception:
+        # Should be AttributeError or io.UnsupportedOperation, but catching
+        # all exceptions here won't hurt
+        return False
+
+
+def _reader_to_websocket(reader, ws, encoding, cancel_reader=None, bufsize=128*1024):
+    """Read reader through to EOF and send each chunk read to the websocket."""
+    while True:
+        if cancel_reader is not None:
+            result = select.select([cancel_reader, reader], [], [])
+            if cancel_reader in result[0]:
+                os.close(cancel_reader)
+                break
+
+        chunk = reader.read(bufsize)
+        if not chunk:
+            break
+        if isinstance(chunk, str):
+            chunk = chunk.encode(encoding)
+        ws.send_binary(chunk)
+
```

Comment:
where does the cancel_reader get closed in the case the reading completed successfully?

Reply:
Good call. I've fixed this now.